### PR TITLE
Fix lightweight delete after drop of projection

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix lightweight deletes after drop of projection. Lightweight deletes do not work in presence of projection but previously they could fail even after drop of projection.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 2471b032ab7a504d1997e9d3681bf97f0564273d with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52517/2471b032ab7a504d1997e9d3681bf97f0564273d/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -5693,7 +5693,8 @@ bool MergeTreeData::supportsLightweightDelete() const
     auto lock = lockParts();
     for (const auto & part : data_parts_by_info)
     {
-        if (!part->supportLightweightDeleteMutate())
+        if (part->getState() == MergeTreeDataPartState::Active
```
Comment (davenger) [inline src/Storages/MergeTree/MergeTreeData.cpp:5696]: I think we should also return false for these states (not 100% sure though)
```
    Temporary,       /// the part is generating now, it is not in data_parts list
    PreActive,       /// the part is in data_parts, but not used for SELECTs
```

Maybe it's better to inverse the check and explicitly skip parts in Deleting and Outdated states like this:

```
     for (const auto & part : data_parts_by_info)
     {
        if (part->getState() == DataPartState::Outdated ||
            part->getState() == DataPartState::Deleting)
            continue;

         if (!part->supportLightweightDeleteMutate())
             return false;
     }
```

Comment (CurtizJ) [reply src/Storages/MergeTree/MergeTreeData.cpp:5696]: I think, you are right.

**Comment by CurtizJ** [comment]
I also found that this sequence of queries fails on materialization of projection in an obscure way, but let's fix it in a separate PR:

```sql
SET mutations_sync = 2;

CREATE TABLE t_projections_lwd (a UInt32, b UInt32, PROJECTION p ()) ENGINE = MergeTree;

INSERT INTO t_projections_lwd SELECT number, number FROM numbers(100);

-- LWD works
DELETE FROM t_projections_lwd WHERE a = 0;

-- add projection
ALTER TABLE t_projections_lwd ADD PROJECTION p_t_projections_lwd ();
ALTER TABLE t_projections_lwd MATERIALIZE PROJECTION p_t_projections_lwd;
```

**Review by davenger — APPROVED** [review]

**Pull request merged**